cirq/protocols/circuit_diagram_info.py

```diff
@@ -43,23 +45,30 @@ def __init__(self,
                 (unless it's equal to 1). For example, the square root of X gate
                 has a text diagram exponent of 0.5 and symbol of 'X' so it is
                 drawn as 'X^0.5'.
-            connected: Whether or not to draw a line connecting the qubits.
+            vconnected: Whether or not to draw a vertical line connecting the
```

Comment:
I don't think this is the right way to go about specifying more complicated drawings. I think we want to have options that are more along the lines of opting into particular functionality as opposed to opting out.

For example, in text diagrams we render disconnected swaps as little Xs joined by vertical lines. That's unlikely to change. Does that mean that the swap gate needs to understand the layout of the qubits in order to set the vconnected field? What if we show the connecting line in text diagrams but not in latex diagrams?

Reply:
`vconnected` is just a renaming of `connected` to distinguish it from the new `hconnected`. All of the non-qcircuit text diagram code works exactly as before, setting `vconnected = hconnected = True` by default. For the qcircuit, `vconnected=False` only for gates that are to be drawn using `\multigate` (a feature that predates this PR). `hconnected` serves an analogous purpose for the swaps. In particular, this PR as-is "show[s] the connecting line in text diagrams but not in latex diagrams". 

Could you say more about your opt-in vs. opt-out distinction? It seems to me that the "functionality" here could be "include the line" just as well as "don't include the line".

Reply:
This is certainly a straightforward way to communicate to the qcircuit drawer "please include \qwx or \qw or both or neither". But does it also apply to the text diagrams and to hypothetical 3d diagrams?

I think what I'm getting at is that we want a gate to have the ability to say "don't you dare *touch* my area in diagrams of style X. I will handle everything there.". I don't want to keep hacking on individual pieces to make individual gates work. More and more we're workaround around the diagramming system being limited in what it was designed to do. We need to redesign the system before our code is nothing but workarounds. This particular workaround isn't the problem in and of itself, it's just the straw breaking the camel's back.

Reply:
That's fair enough. The fact that `TextDiagramDrawer` supports both the text diagrams and qcircuit diagrams is both nice in that it allows code re-use and not nice in that it seems forced when the logic of the two is slightly different.

I'm happy to help out with a more thorough overhaul, and can think more about what that should look like. I opened #1245 to host a discussion on the subject. It would help if you could enumerate the issues you see there. Could we do this one last workaround in the meantime?